### PR TITLE
Email epic

### PR DESCRIPTION
**Pull request by AbbyJonesDev**
It's been a long time since we sketched out the plan for this, but here's a Google Doc with an updated outline of how things are set up:
https://docs.google.com/document/d/1wAru-qnW0XJZlOf7wBnnqI85LyQZJ72li1NgDQLSk6A/edit?usp=sharing

And a look at what's included now...  (Note - I've merged from master, so not all of the changed files are actually my files.  I'll do another merge and run the test suite one more time after I finish some refactoring.)

* MailCampaignHelper - a module that can be dropped into other mailers and make it easy to utilize Sendgrid's API and record sent emails.

* NudgeMailer - for nudging folks who haven't logged in for awhile...  (Also serves as a nice blueprint for future email campaigns.)  Currently has just one method - inactive_one_month

* Footer - contains unsubscribe link and contact information (per the CAN SPAM act).  Can be added to any email layout.
app/views/shared/_mailer_footer.html.erb

* Link in the footer leads to a generic unsubscribe page where users can unsubscribe from selected lists or all lists.  (Unsubscribing from all also sets flag on that user.)

* User profile page contains a link to update email preferences.  From that page, users can add or remove unsubscriptions.  Unsubscribe_all flag is added or removed as appropriate.

### CODE REVIEW

**Comment by eriktrautman** [comment]
The oldest, and still largest, outstanding PR :)

File: app/helpers/mail_campaign_helper.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      recipient_addresses << user.email
+    end
+    return recipients, recipient_addresses
+  end
+
+  def build_header(recipients)
+    header = {
+      "to" => recipients
+    }.to_json
+    headers["X-SMTPAPI"] = header
```
Comment (AbbyJonesDev) [inline app/helpers/mail_campaign_helper.rb]: This sets the header that Sendgrid will parse to deliver the email to multiple recipients

File: app/helpers/mail_campaign_helper.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      next if user.unsubscriptions.where(
+        email_campaign_category_id: category_id) != []
+      # Skip if user has already gotten this email
+      next if user.sent_emails.where(
+        email_campaign_id: campaign.id) != []
+      # Still here? Add email to recipients array
+      recipients << user
+      recipient_addresses << user.email
+    end
+    return recipients, recipient_addresses
```
Comment (AbbyJonesDev) [inline app/helpers/mail_campaign_helper.rb]: The recipient email addresses will be used to build the smtpapi header and send the emails.
The recipients (User objects) will be needed to record sent emails.
The filter method iterates through possible users one time and builds both lists to save more time/database calls later on.

File: app/models/unsubscription.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    ecc_list = []
+    category_names.each do |c|
+      ecc_list << EmailCampaignCategory.find_by_name(c)
+    end
+    ecc_list.each do |c|
+      Unsubscription.create(user_id: user.id, email_campaign_category_id: c.id)
+    end
+  end
+
+  def self.category_names(unsubscriptions)
```
Comment (AbbyJonesDev) [inline app/models/unsubscription.rb]: This method is only used in tests...  I'll be exorcising it.

**Comment by AbbyJonesDev** [comment]
Refactored and ready for review.

File: app/controllers/contacts_controller.rb [hunk 4]
```diff
@@ -1,37 +1,33 @@
 class ContactsController < ApplicationController
-  before_filter :authenticate_user!
+  before_action :authenticate_user!
+  before_action :set_subject
```
Comment (herminiotorres) [inline app/controllers/contacts_controller.rb]: Best only set to create action, it now only uses the `@subject`

File: app/assets/stylesheets/users.css.scss [hunk 5]
```diff
@@ -1,3 +1,26 @@
...
+@media(max-width: 480px){
+  .btn-email-pref{
+    margin-left: 0;
+  }
+}
+
+.btn-email-pref:hover {
+  background: #303030;
+  color: #fff;
+}
```
Comment (herminiotorres) [inline app/assets/stylesheets/users.css.scss]: you have a set up your text editor to include a new blank line at the end the file when you saving.

File: app/assets/stylesheets/unsubscriptions.css.scss [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+.unsubscriptions-controller {
+  width: 70%;
+  margin: 0 15%;
+}
+
+.email-categories {
+  padding: 10px;
+  // border: solid 1px #444;
+  margin: 5px 5px;
+}
```
Comment (herminiotorres) [inline app/assets/stylesheets/unsubscriptions.css.scss]: you have a set up your text editor to include a new blank line at the end the file when you saving.

File: app/controllers/contacts_controller.rb [hunk 7]
```diff
@@ -1,37 +1,33 @@
...
       flash.now[:error] = "Please fill out the message form"
       render :new
     end
-
   end
 
-  def default_subject
-    "You have a new message from #{current_user.username} of The Odin Project"
+  private
+  def set_subject
```
Comment (herminiotorres) [inline app/controllers/contacts_controller.rb]: you will put a new line between private and the method?

File: app/controllers/contacts_controller.rb [hunk 8]
```diff
@@ -1,37 +1,33 @@
...
-    end
-    @user = user
+    @user = User.find params[:user_id]
+    redirect_to root_path if @user.nil?
   end
 
   def create
-    @user = User.find_by_id(params[:user_id])
+    @user = User.find params[:user_id]
     @message = Message.new({
```
Comment (herminiotorres) [inline app/controllers/contacts_controller.rb]: if it is using the version of Ruby 1.9 or larger, better change this hash structure.
e.g.:
```ruby
Message.new({
  body: params[:body].html_safe,
  ...
})
```

File: app/controllers/contacts_controller.rb [hunk 9]
```diff
@@ -1,37 +1,33 @@
...
         :body => params[:body].html_safe,
```
Comment (herminiotorres) [inline app/controllers/contacts_controller.rb]: You could use `.fetch` in this params? I think is the best way.

File: Gemfile [hunk 10]
```diff
@@ -34,10 +34,10 @@ group :development, :test do
   gem 'fakeweb', '~> 1.3.0'
   gem 'capybara',                   '<= 2.0.2'
   gem 'capybara-webkit', '~> 1.1.1'
+  gem 'jazz_hands'
```
Comment (herminiotorres) [inline Gemfile]: Ruby already has a great debugger, why use `jazz_hands`?

You know this gem has increased the time to load the rails?

File: app/controllers/unsubscriptions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    @categories = EmailCampaignCategory.list
+  end
+
+  def create
+    @user = User.find_by_email(params[:email_address])
+    @categories = params[:categories]
+    verify_user; return if performed?
+    confirm_categories_are_checked; return if performed?
+    create_unsubscriptions
+    check_unsubscribe_all; return if performed?
```
Comment (herminiotorres) [inline app/controllers/unsubscriptions_controller.rb]: `check_unsubscribe_all return if performed? `

**Pull request closed without merging**